_posts/2017/2017-03-10-creating-a-sales-analytics-app-with-ruby-on-rails-2.md

```diff
@@ -0,0 +1,84 @@
+
+Our first approach was to add all the products, customers and orders to our database.
+For this, we were using [sidekiq](http://sidekiq.org/) to send the syncing tasks to the background.
+Sidekiq is fast and really customizable, but even with all the flexibility on the world, if the data isn't there, it can't be used.
+
+At the time of writing this, it seems obvious that shops aren't made of two demo customers and three orders with one product each.
+But when we where starting the project, that was our test data and this answer to the synchronization problem seemed to be, at least, a not bad one.
+Turns out, it wasn't.
+
+When we got some real data and we were left with a white graph for the first couple of hours (remember the hourly rate limit) it became clear that our answer wasn't the solution.
+So we turned to the pragmatical solutions, only getting the data we needed.
+
+The bad thing about getting the data at the moment is that you have to persist it and do the calculations **and** also show something on the screen.
+So, with this mentality still on mind, we decided to fetch the orders first.
+This way we only retrieve the customers and products we need, if needed.
+We save database space, API calls and we show the data in a chronological order, so the merchant can at least, even in the worst scenarios, see the latest data.
+
+So for now, the app checks every hour if there are any new orders to download the latest data or just schedule itself to check for that the next hour.
+We also decided that we were only going to get data not older that two years, so we don't get overloaded with old data and because the main use case for the app is to be able to check and get reports and analysis closer in time data, so it's more relevant.
+
+Our last issue was with encryption.
+By default, Rails encrypts passwords and we were already encrypting the `access_token`, but we got the request to encrypt the email addresses also.
+The problem with this is that the library we were using for user management, the easy and simple `Devise`, didn't support that.
+Changing to username for the log in process was the first thing it came into our minds.
+
+This is probably the easiest solution but also, one that changed how the app worked.
+Now the merchants needed to think about a username and then, remember it.
+Our goal was to make the app the most accessible as possible and adding extra fields to remember was not the way to do it.
+
+The other answer that came to my mind was *just* to adapt the log in system to use encrypted emails.
+At that moment we didn't fully realize what was the scope of that task we were creating, the easy and simple Devise was about to become a pain all over the app in ways that we couldn't even dream of figuring it out.
+
+I started creating two methods for auto-encrypt and auto-decrypt any database field via Ruby Magic (AKA Metaprogramming).
+This way we could add encryption easily where needed. But then Devise said **no**.
```

Comment:
No, sorry. I rewrote just the first sentence, the second was fine. Please see below:

Doing so, we were able to add encryption easily where required.
But then Devise said **no**.